Allow field mappers to retrieve fields from source.

jtibshirani opened the pull request:
This PR adds new method `FieldMapper#lookupValues(SourceLookup)` that extracts and parses the source values. This lets us return values like numbers and dates in a consistent format, and also handle special data types like `constant_keyword`. The `lookupValues` method calls into `parseSourceValue`, which mappers can override to specify how values should be parsed.

Older draft PR: #56473 

Relates to #55363.

jtibshirani committed:
Introduce a method FieldMapper#lookupValues to retrieve fields from source.

elasticmachine commented:
Pinging @elastic/es-search (:Search/Mapping)

jtibshirani commented:
As part of this PR, we need to decide on the 'standard format' for each field type. I'll work to collect feedback/ discussion on this point. Here's the current proposal:
* `boolean`: parse into boolean
* `binary`: no change to value
* `completion`: no change to value
* `constant_keyword`: don't consult source, return the constant value
* `date`: parse to long millis based on mapping format
* `date_nanos`: parse to long nanos based on mapping format
* `ip`: no change to value
* `keyword`: convert value to string, do not apply normalization. 
* numeric types (`integer`, `float`, etc.): parse value to the right type, for example for a `long` field "42.3" becomes 42
* range types (`integer_range`, etc.): parse each bound to the right type, this includes both numbers and dates
* `rank_feature`: convert value to float
* `text`, `annotated_text`: convert value to string

Some notable examples where we don't parse at all:
* `histogram`, `rank_features`, `vector`: no change to value, don't attempt to convert each component into a number. For me the complexity of digging through and standardizing these structured values outweighed the benefit.
* `murmur3`, `token_count`: no change to value, don't transform input. It didn't seem practical to always recompute a hash/ token count, and I couldn't see an important use case for returning these values.

We never return a field that has been dropped through `ignore_malformed`. This way, the values returned for particular field will always have the same type and format. The idea is that the value is ignored in all high-level parts of the search request, including the query, aggs, and field retrieval.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:
```diff
@@ -187,6 +187,11 @@ public KeywordFieldMapper build(BuilderContext context) {
     }
 
+    @Override
+    protected Object parseSourceValue(Object value) {
```
nit: I was wondering if we should use covariance here and anywhere we return Strings, like we do for other types.

jtibshirani replied:
👍

jtibshirani committed:
Ensure parseSourceValue returns strings when possible.

nik9000 commented:
> `date`: parse to long millis based on mapping format

This is one place where `format` is pretty important, I think. We should probably apply whatever the default "format" is for this field. Though that can totally be a follow up.

Also in a follow up, folks should be able to specify the format, I think. That way folks can say things like "give this is millis-since-epoch even if the default format is ISO8601".

I'm not sure if we need time zone support too. I know the aggs have it, but that feels like a big much at this point.

nik9000 commented:
> `ip`: no change to value

Should this canonicalize the IP?

jtibshirani commented on server/src/main/java/org/elasticsearch/index/mapper/RangeType.java:
```diff
@@ -619,14 +629,17 @@ public Object parseTo(RangeFieldMapper.RangeFieldType fieldType, XContentParser
-        Object lower = from == null ? minValue() : parse(from, false);
-        Object upper = to == null ? maxValue() : parse(to, false);
+        Object lower = from == null ? minValue() : parseValue(from, false, dateMathParser);
+        Object upper = to == null ? maxValue() :parseValue(to, false, dateMathParser);
```
Oops, thanks. I actually opened https://github.com/elastic/elasticsearch/pull/56976 to refactor `RangeType` separately -- once that's merged I'll rebase this PR on top of it.

jtibshirani commented:
> This is one place where format is pretty important, I think. We should probably apply whatever the default "format" is for this field. Though that can totally be a follow up.

I'll handle this now, I do think it fits best within this PR!

> Also in a follow up, folks should be able to specify the format, I think. That way folks can say things like "give this is millis-since-epoch even if the default format is ISO8601".

+1, this is tracked in the meta-issue as a follow-up (#55363).

> Should this canonicalize the IP?

Sorry if I'm missing something obvious, what's an example of the canonicalization you have in mind ?

nik9000 commented:
> Sorry if I'm missing something obvious, what's an example of the canonicalization you have in mind ?

I believe round tripping IPs through doc values would apply [this](https://en.wikipedia.org/wiki/IPv6_address#Representation).

nik9000 commented:
> I believe round tripping IPs through doc values would apply [this](https://en.wikipedia.org/wiki/IPv6_address#Representation).

I think it might be as simple as parsing and `toString`ing to make that happen though.

jtibshirani committed:
Convert murmur3 and token_count values to strings.

jtibshirani committed:
Format dates based on the mapping format.

jtibshirani commented:
A heads up that I rebased and force-pushed to incorporate a refactor (#56976).

astefan commented on server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:
```diff
@@ -187,6 +187,11 @@ public KeywordFieldMapper build(BuilderContext context) {
 
+    @Override
+    protected String parseSourceValue(Object value) {
+        return value.toString();
```
Keywords shouldn't also consider `ignore_above` and the extraction from `_source` to happen with this value in mind?

jtibshirani replied:
I'm planning to tackle `ignore_above` (and likely `null_value` too) in a follow-up PR.

astefan commented on x-pack/plugin/mapper-constant-keyword/src/test/java/org/elasticsearch/xpack/constantkeyword/mapper/ConstantKeywordFieldMapperTests.java:
```diff
@@ -127,4 +130,27 @@ public void testMeta() throws Exception {
+    public void testLookupValues() throws Exception {
+        IndexService indexService = createIndex("test");
+        String mapping = Strings.toString(XContentFactory.jsonBuilder().startObject().startObject("_doc")
+            .startObject("properties").startObject("field").field("type", "constant_keyword")
```
Really minor: all the indentations for the lines containing method calls in `testLookupValues` are inconsistent compared to the other methods in this class.

jtibshirani replied:
👍

astefan commented on x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java:
```diff
@@ -551,6 +551,11 @@ protected void parseCreateField(ParseContext context) throws IOException {
 
+    @Override
+    protected String parseSourceValue(Object value) {
+        return value.toString();
```
Same comment as the one for `keyword` field: shouldn't `ignore_above` be considered when parsing the source?

astefan reviewed (approved):
LGTM.
Left few minor comments.

jtibshirani committed:
Make sure to format dates within range types.

jtibshirani committed:
Fix indentation.

jtibshirani committed:
Standardize IP fields.

jtibshirani commented:
Thanks @astefan for the review!

@nik9000 it would be great to get your review as well. Here are the changes since last time:
* We now format dates using the format in the mapping. This applies to `date`, `date_nanos`, and `date_range`.
* IPs are standardized using the utilities in `InetAddresses`. I don't have a strong intuition here, but it felt like a good choice since it matches what's returned in `docvalue_fields`.

jtibshirani commented:
@elasticmachine run elasticsearch-ci/bwc

nik9000 reviewed (approved):
LGTM! Sorry for the delay.
